merlin_ai/ai_classes.py

```diff
@@ -301,3 +311,77 @@ def _generate_prompt(
                 {"role": "user", "content": f"The text to classify:\n{value}"},
             ],
         )
+
+
+class OpenAIEnumExplained(OpenAIEnum):
+    """
+    OpenAI-based AI Enum with explanation
+    """
+
+    def __init__(self, data_class: Type, model_settings: Optional[dict] = None):
+        super().__init__(data_class, model_settings)
+        fields = [("category", data_class,
+                   field(metadata={"description": data_class.__doc__} if data_class.__doc__ else None)),
+                  ("explanation", str,
+                   field(metadata={"description": "Explain your categorization in a short and concise manner."}))
+                  ]
+        data_class_wrapper = dataclasses.make_dataclass(f"{data_class.__name__}_wrapper", fields)
+        if data_class.__doc__:
+            data_class_wrapper.__doc__ = data_class.__doc__
+        self._data_class_wrapper = data_class_wrapper
+
+    def __str__(self):
+        return f"OpenAIEnumExplained: {self._data_class.__name__}"
+
+    def create_instance_from_response(self, llm_response):
+        content = json.loads(llm_response.choices[0].message.function_call.arguments)
+        enum_options = self._get_enum_options()
+
+        for option in enum_options:
+            if option.name == content["category"]:
+                return option
+
+        raise RuntimeError(f"LLM returned invalid value {content['category']}")
+
+    def _get_llm_settings(
+        self, function_call_model_settings: Optional[dict] = None
+    ) -> dict:
+        return self._get_base_llm_settings(function_call_model_settings)
+
+    def _generate_prompt(
+        self, value: str, model_settings: dict, instruction: Optional[str] = None
+    ) -> PromptBase:
+        function_name = "format_response"
+        model_settings["functions"] = [
+            OpenAIModel.generate_function_call_object(self._data_class_wrapper)
+        ]
+        model_settings["function_call"] = {"name": function_name}
+
+        enum_options = self._get_enum_options()
+        system_prompt = "You are an expert classifier that always chooses correctly\n\n"
+        if not instruction:
+            instruction = self._data_class.__doc__.strip()
+        if instruction != "An enumeration.":
+            system_prompt += f"Also note that:\n{instruction}\n\n"
+
+        system_prompt += (
+            "The user will provide text to classify, you will use your expertise "
```

Comment:
If you require LLM to call your function, you must also include this explicitly in prompt (see AIModel system prompt)

Reply:
Good catch! This works in testing as is, but cannot hurt to instruct is explicitly. Will add.